### PR TITLE
Add explicit interface methods

### PR DESCRIPTION
**Pull request by apaszke**
It is often useful to define type -> value mappings, and the standard
way to do that is through interfaces (type classes). However, because
all methods so far had implicit type parameters, any attempt to
associate e.g. an integer with a type was difficult: without explicit
type annotations it often ended up being ambiguous.

This patch allows specifying interface parameters between each method
name and the colon that begins its type annotations, with the mentioned
parameters becoming explicit type parameters of the generated methods.
This lets us remove the awkward `TypeVehicle` abstraction from prelude
and in the future should make it possible to define associated types.
For example, a `Manifold a` interface could have a method declared as
`TangentSpace a : Type`, which would later make it possible to mention
`TangentSpace MyDataType` in types.

### CODE REVIEW

**Comment by oxinabox** [comment]
would this unblock #343 ?

**Comment by apaszke** [comment]
Not yet, but almost! This PR + #659 + ~50 LOC more are what we need to unblock it!

**Comment by danieldjohnson** [comment]
I vaguely remember having a conversation where we considered spelling this as
```
interface Manifold a
  TangentSpace a : Type
```
as opposed to
```
interface Manifold a
  explicit TangentSpace : Type
```
Do you think that's worth doing?

**Comment by apaszke** [comment]
Oh, that's a nice one! It also has the benefit that you can (1) list a subset of type-class binders (2) permute them. My only concern would be that
```
someMethod a : a -> Int
```
might be a little hard to parse, because the `a` on the lhs implies an argument of type `Type` and the one on rhs of `:` implies an argument of type `a`. But I guess if someone has a method that does take an instance then they won't declare it with an explicit type parameter?

**Comment by apaszke** [comment]
I switched the implementation to use what @danieldjohnson has suggested. It has a temporary restriction that you either specify all or no type parameters (and without any permutations), but we can relax it in the future.
